### PR TITLE
feat(limits): propagate generate limits

### PR DESCRIPTION
**Pull request by Tomas2D**
---
## Status
**READY**

## Description

`generate_async` now orchestrates the generation tasks by syncing its limits state with the API rather than having a hard-coded value. Also, the graceful shutdown of the app should be faster as there will be fewer pending tasks.

## Impacted Areas in Library
Model - `generate_async` method

## Which issue(s) does this pull-request fix?
Closes: #147

## Any special notes for your reviewer?

Sleep timeouts - still not sure about ideal values (sleep and caching)

---

## Checklist
- [ ] Automated tests exist
- [ ] Updated Package Requirements (if required, and with maintainers' approval)
- [ ] Local unit tests performed
- [ ] Documentation exists [link]()
- [ ] Local pre-commit hooks performed
- [ ] Desired commit message set as PR title and description set above
- [ ] Link to relevant GitHub issue provided

### CODE REVIEW

File: src/genai/model.py [hunk 1]
```diff
@@ -196,31 +193,31 @@ def generate_async(
...
+            ) as generator:
                 for response in tqdm(
-                    asynchelper.generate_response(),
+                    iterable=generator.generate_response(),
                     total=len(prompts),
                     desc="Progress",
                     unit=" inputs",
                     disable=hide_progressbar,
                 ):
+                    print("Helloo")
```
Comment (David-Kristek) [inline src/genai/model.py:217]:
```suggestion
```

File: examples/user/async_greetings.py [hunk 2]
```diff
@@ -29,13 +29,19 @@
...
 num_of_greetings = len(lots_of_greetings)
 num_said_greetings = 0
 greeting1 = "Hello! How are you?"
 
 # yields batch of results that are produced asynchronously and in parallel
-for result in model.generate_async(lots_of_greetings):
+for result in model.generate_async(
+    lots_of_greetings,
+    ordered=True,
+    # throw_on_error=True,
```
Comment (David-Kristek) [inline examples/user/async_greetings.py:41]: Is the comment wanted?

**Review by David-Kristek — CHANGES_REQUESTED** [review]

File: src/genai/services/async_generator.py [hunk 3]
```diff
@@ -124,66 +125,76 @@ async def _get_response_json(self, model, inputs, params, options):
...
+            logger.error(
+                "Exception raised in callback : {}, response = {}, inputs = {}".format(str(e), response, inputs)
+            )
 
     async def _schedule_requests(self):
+        async def get_limits():
+            if self.fn == "tokenize":
+                return len(self.prompts)
+
+            limits = self.service.generate_limits()
```
Comment (David-Kristek) [inline src/genai/services/async_generator.py:170]: I think that this function should be mocked in testing in test_concurrent.py. - test is failing

**Review by David-Kristek — CHANGES_REQUESTED** [review]

File: src/genai/extensions/langchain/llm.py [hunk 4]
```diff
@@ -156,6 +161,17 @@ def _generate(
...
+    async def _agenerate(
+        self,
+        prompts: List[str],
+        stop: Optional[List[str]] = None,
+        run_manager: Optional[AsyncCallbackManagerForLLMRun] = None,
+        **kwargs: Any,
+    ) -> LLMResult:
+        return await asyncio.get_running_loop().run_in_executor(
+            None, partial(self._generate, prompts, stop, run_manager, **kwargs)
+        )
```
Comment (Tomas2D) [inline src/genai/extensions/langchain/llm.py:173]: This part has been updated in a newer version of LangChain, but we want to support the old behaviour.

**Review by David-Kristek — APPROVED** [review]

**Pull request merged**